File: src/Cpp/18_Operacje_na_plikach/Zad9.cpp
#include <cassert>
#include <experimental/filesystem>
#include <fstream>
#include <string>

namespace filesys = std::experimental::filesystem;

/*
Otrzymujesz napis reprezentujacy sciezke folderu. Usun
wszystkie pliki o rozmiarach wiekszych niz 10 kB znajdujace
sie w podanym folderze oraz jego podfolderach.
*/

int rozmiarPliku(const std::string &sciezka) {
  return filesys::file_size(filesys::path(sciezka));
}

void usunPliki(const std::string &sciezka) {

  for (const auto plik : filesys::directory_iterator(sciezka)) {
    if (rozmiarPliku(plik.path()) > 10240)
      filesys::remove(filesys::path(plik.path()));
  }
}

void testUsunPliki() {

  // stworz folder testowy
  std::string sciezka = "test";
  filesys::create_directory(filesys::path(sciezka));

  // dodaj plik do folderu
  std::string sciezkaPliku = "test/plik.txt";
  std::ofstream plik(sciezkaPliku);
  for (int i = 0; i < 100000; i++)
    plik << "test\n";
  plik.close();

  // usun pliki z testowego folderu
  usunPliki(sciezka);

  // sprawdz czy folder test jest pusty
  assert(filesys::is_empty(filesys::path(sciezka)));

  // usun folder test
  filesys::remove_all(filesys::path(sciezka));
}

int main() {

  testUsunPliki();

  return 0;
}

File: src/js/18_operacje_na_plikach/zad9.js
/*
Otrzymujesz napis reprezentujacy sciezke folderu. Usun
wszystkie pliki o rozmiarach wiekszych niz 10 kB znajdujace
sie w podanym folderze oraz jego podfolderach.
*/
const fs = require("fs").promises;
const path = require("path");

usunPliki = async (sciezka, rozmiar) => {
  const pliki = await fs.readdir(sciezka);
  const plikiTxt = pliki.filter((plik) => plik.endsWith(".txt"));
  for (let i = 0; i < plikiTxt.length; i++) {
    const plik = plikiTxt[i];
    const sciezkaPliku = path.join(sciezka, plik);
    const rozmiarPliku = (await fs.stat(sciezkaPliku)).size;
    if (rozmiarPliku > rozmiar) {
      await fs.unlink(sciezkaPliku);
    }
  }
};

test1 = async () => {
  const sciezkaFolderu = "temp_dir";
  const sciezkikiPlikow = ["lista.txt", "test.txt"];
  const tekst = "przykladowy tekst.\n";

  await fs.mkdir(sciezkaFolderu);

  for (const plik of sciezkikiPlikow) {
    const sciezkaPliku = path.join(sciezkaFolderu, plik);
    await fs.writeFile(sciezkaPliku, tekst.repeat(10000));
  }

  await usunPliki(sciezkaFolderu, 10 * 1024);

  const daneFolderu = await fs.readdir(sciezkaFolderu);
  if (daneFolderu.length !== 0) {
    throw new Error(`Assertion error line 180: ${daneFolderu.length} != 0`);
  }

  await fs.rmdir(sciezkaFolderu, {
    recursive: true,
  });
};

main = async () => {
  await test1();
};

main();
